### PR TITLE
implement ssl-skip-verify to forward to self-signed-certificates

### PR DESCRIPTION
**Pull request by lorenz-haenggi-jarowa**
we are using internally also SSL between our servers (nginx and Service fabric / load balancer) but these are self-signed-certiticates. Today clammit does not allow it

`
2021/11/04 23:18:04 Interceptor passed this request
2021/11/04 23:18:04 Forwarding to https://10.57.2.4:8001
2021/11/04 23:18:04 Failed to forward request: 'post' "https://10.57.2.4:8001/api/Files?api-version=1.0": x509: certificate signed by unknown authority
`

the PR allows to configure "ssl-skip-verify" (true, false). 

#
# If true, will skip SSL validation forwarding connection to use self-signed certitifates
# default = false
#
#ssl-skip-verify = false

`
2021/11/04 23:22:26 Interceptor passed this request
2021/11/04 23:22:26 Forwarding to https://10.57.2.4:8001
2021/11/04 23:22:26 Request forwarded, response 200 OK
`

- test cases ran
- test cases implemented
- documentation adapted
- (some build documentation adapted)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,6 +3,9 @@ all:	gets application
 linux:
 	GOOS=linux GOARCH=amd64 make application
 
+linux-win:
+	set GOOS=linux && set GOARCH=amd64 && make application
+
```
Comment (vjt) [inline Makefile:8]: For consistency, can you please use bash syntax, avoiding `set` and `&&`?

File: README.md [hunk 2]
```diff
@@ -144,6 +144,7 @@ content-memory-threshold | (Optional) Maximum payload size to keep in RAM. Large
 log-file                 | (Optional) The clammit log file, if ommitted will log to stdout
 test-pages               | (Optional) If true, clammit will also offer up a page to perform test uploads
 debug                    | (Optional) If true, more things will be logged
+ssl-skip-verify          | (Optional) If true, will skip SSL validation forwarding connection to use self-signed certitifates
```
Comment (vjt) [inline README.md:147]: I would reword this as:

> if true, it will not perform TLS certificate verification, allowing connections to servers using self-signed or otherwise untrusted certificates. Use with care, it is always preferable to perform certificate verification.

File: clammit.cfg.example [hunk 3]
```diff
@@ -31,3 +31,9 @@ log-file        = log/clammit.log
 # the virus scanning
 #
 #test-pages      = true
+
+#
+# If true, will skip SSL validation forwarding connection to use self-signed certitifates
+# default = false
```
Comment (vjt) [inline clammit.cfg.example:37]: Also here I would reword the description as:

> if true, it will not perform TLS certificate verification, allowing connections to servers using self-signed or otherwise untrusted certificates. Use with care, it is always preferable to perform certificate verification.

File: forwarder/forwarder.go [hunk 4]
```diff
@@ -236,6 +239,11 @@ func (f *Forwarder) getClient(req *http.Request) (*http.Client, *url.URL) {
 		}, url
 	} else {
 		f.logger.Printf("Forwarding to %s", applicationURL.String())
-		return &http.Client{}, url
+		// allow for
+		// https://stackoverflow.com/questions/12122159/how-to-do-a-https-request-with-bad-certificate
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: f.sslSkipVerify},
```
Comment (vjt) [inline forwarder/forwarder.go:245]: This looks OK as today we only have a single option. For the future, we’ll want to pass the tls config object in the forwarder struct directly, that’ll allow us to drive any tls config options from clammit configuration without adding more fields to the Forwarder struct.

**Review by vjt — COMMENTED** [review]
Thank you for this. I have mixed feelings about this change because:

* I think that TLS verification is a good thing and one should always employ it
* I am not convinced on how the sslskipverify option is passed around

So I summon @jblackman for his feedback :-)

File: forwarder/forwarder.go (hunk 4 above)
Comment (lorenz-haenggi-jarowa) [reply forwarder/forwarder.go:245]: I agree - I was not happy also. But I did not wanted to do a full refactoring ....

File: clammit.cfg.example (hunk 3 above)
Comment (lorenz-haenggi-jarowa) [reply clammit.cfg.example:37]: I like your very precise documentation

File: README.md (hunk 2 above)
Comment (lorenz-haenggi-jarowa) [reply README.md:147]: I like your very precise documentation

File: Makefile (hunk 1 above)
Comment (lorenz-haenggi-jarowa) [reply Makefile:8]: ok - then I remove it again

**Comment by butsjoh** [comment]
Any change on getting this merged?

**Comment by butsjoh** [comment]
Also another question/observation: Why is it not possible that self signed certs that have been added to the os level are considered valid? I am not familiar with the go ecosystem but i would expect that a cert that is trusted on the system level would be working. We tried this together with https://github.com/maxsivkov/clammit-docker (and after adding the necessary cert files) and when we curl inside the container it works but when clammit forwards the request it keeps complaining about invalid cert.
